### ISSUE TITLE
problem using groovy-lint on M1 mac (arm64) - Error unsupported architecture

### ISSUE THREAD

**Issue by s-bernhardt-atos** [nvuillam/vscode-groovy-lint#163]
Hi Team,
I debugged the logs a little and found the probable root cause, why my linter is failing on my new Mac (M1 Pro).

```
2022-08-24T09:55:29.954Z vscode-groovy-lint Request execute npm-groovy-lint for file:///Users/.../pipelines/Jenkinsfiles/deploy-test/Jenkinsfile.install-microk8s-cluster with options {"displayErrorsEvenIfDocumentClosed":false}
2022-08-24T09:55:29.955Z vscode-groovy-lint Reset diagnostics for file:///Users/.../pipelines/Jenkinsfiles/deploy-test/Jenkinsfile.install-microk8s-cluster 
Start analyzing file:///Users/.../pipelines/Jenkinsfiles/deploy-test/Jenkinsfile.install-microk8s-cluster 
Java jre or jdk between 8 and 14 is required 
Installing Java jre 14 in /Users/stefan/.java-caller...
**VsCode Groovy Lint error: Unsupported architecture**
Error: Unsupported architecture
    at Object.install (/Users/stefan/.vscode/extensions/nicolasvuillamy.vscode-groovy-lint-2.0.0/server/node_modules/njre/lib/install.js:204:32)
    at JavaCaller.manageJavaInstall (/Users/stefan/.vscode/extensions/nicolasvuillamy.vscode-groovy-lint-2.0.0/server/node_modules/java-caller/lib/java-caller.js:254:43)
    at async JavaCaller.run (/Users/stefan/.vscode/extensions/nicolasvuillamy.vscode-groovy-lint-2.0.0/server/node_modules/java-caller/lib/java-caller.js:86:13)
    at async CodeNarcCaller.startCodeNarcServer (/Users/stefan/.vscode/extensions/nicolasvuillamy.vscode-groovy-lint-2.0.0/server/node_modules/npm-groovy-lint/lib/codenarc-caller.js:266:29)
    at async CodeNarcCaller.callCodeNarcServer (/Users/stefan/.vscode/extensions/nicolasvuillamy.vscode-groovy-lint-2.0.0/server/node_modules/npm-groovy-lint/lib/codenarc-caller.js:102:22)
    at async NpmGroovyLint.callCodeNarc (/Users/stefan/.vscode/extensions/nicolasvuillamy.vscode-groovy-lint-2.0.0/server/node_modules/npm-groovy-lint/lib/groovy-lint.js:237:32)
    at async NpmGroovyLint.run (/Users/stefan/.vscode/extensions/nicolasvuillamy.vscode-groovy-lint-2.0.0/server/node_modules/npm-groovy-lint/lib/groovy-lint.js:66:13)
```

I installed 'jre-8u341-macosx-x64.dmg', additional to macOS Java, but no luck so far.
```bash
➜  ~ /Library/Internet\ Plug-Ins/JavaAppletPlugin.plugin/Contents/Home/bin/java -version
java version "1.8.0_341"
Java(TM) SE Runtime Environment (build 1.8.0_341-b10)
Java HotSpot(TM) 64-Bit Server VM (build 25.341-b10, mixed mode)

➜  ~ /usr/bin/java -version
java version "17.0.2" 2022-01-18 LTS
Java(TM) SE Runtime Environment (build 17.0.2+8-LTS-86)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.2+8-LTS-86, mixed mode, sharing)
```

Did somebody solve this already?

**Comment by s-bernhardt-atos** [comment]
I found in `~/.vscode/extensions/nicolasvuillamy.vscode-groovy-lint-2.0.0/server/node_modules/njre/lib/install.js`
the missing architecture `arm64`.

```js
function install (version = 8, options = {}) {
  const { openjdk_impl = 'hotspot', release = 'latest', type = 'jre' } = options
  options = { ...options, openjdk_impl, release, type }
  let url = 'https://api.adoptopenjdk.net/v2/info/releases/openjdk' + version + '?'

  if (!options.os) {
    switch (process.platform) {
      case 'aix':
        options.os = 'aix'
        break
      case 'darwin':
        options.os = 'mac'
        break
      case 'linux':
        options.os = 'linux'
        break
      case 'sunos':
        options.os = 'solaris'
        break
      case 'win32':
        options.os = 'windows'
        break
      default:
        return Promise.reject(new Error('Unsupported operating system'))
    }
  }
  if (!options.arch) {
    if (/^ppc64|s390x|x32|x64$/g.test(process.arch)) options.arch = process.arch
    else if (process.arch === 'ia32') options.arch = 'x32'
    else return Promise.reject(new Error('Unsupported architecture'))
  }
```

**Comment by nvuillam** [comment]
Indeed... what if you try to install java by yourself ? :)
that way, node-java-caller won't try to install java and it should work directly :)

**Comment by s-bernhardt-atos** [comment]
> Indeed... what if you try to install java by yourself ? :) that way, node-java-caller won't try to install java and it should work directly :)

That's what I tried as well. As you can see above I installed Oracle Java on top.
That didn´t help. Then I tried with `OpenJDK17U-jdk_aarch64_mac_hotspot_17.0.4.1_1.pkg` and removed Oracle Java 1.8.0 before.
Could it be, that there's a max version of 14? Why does it try to install java 14, if I have java already installed (v17.x)

my state is now:
```bash
➜  ~ /usr/bin/java -version
openjdk version "17.0.4.1" 2022-08-12
OpenJDK Runtime Environment Temurin-17.0.4.1+1 (build 17.0.4.1+1)
OpenJDK 64-Bit Server VM Temurin-17.0.4.1+1 (build 17.0.4.1+1, mixed mode)
```

**Comment by nvuillam** [comment]
I'm glad you found a way :)

**Comment by s-bernhardt-atos** [comment]
Thanks for the quick response and the great tool!

**Comment by nfantone** [comment]
Why would this plugin not work with recent JAVA versions? Why is 14 the latest supported one?

**Comment by nvuillam** [comment]
There's indeed no reason, I'll try to arrange that :)

**Comment by mohsen0** [comment]
Can we reopen the issue? I am not sure how to pass JAVA_HOME to the plugin

**Comment by nvuillam** [comment]
I have a PR in progress for some time

https://github.com/nvuillam/npm-groovy-lint/pull/259

But test cases are failing and I have some issues to find time to complete it...

If someone has more availability than me, please feel free to try to complete it, i can't provide a delivery date in my current situation :/

**Comment by NicolasPelletier** [comment]
All good fellow Nicolas. Thanks for replying so fast and linking the PR. Hopefully, somebody more knowledgeable than me can take a look at it.

**Comment by JuanMongeLL** [comment]
> I could get it working with `OpenJDK11U-jdk_aarch64_mac_hotspot_11.0.16.1_1.pkg`, but I had to remove the newer versions. [link](https://github.com/adoptium/temurin11-binaries/releases/download/jdk-11.0.16.1%2B1/OpenJDK11U-jdk_aarch64_mac_hotspot_11.0.16.1_1.pkg)

This is the solution for Mac m2 at least now in July 2023. Kudos for the man for figuring this out.

**Comment by kkreso** [comment]
Tried everything, even with `OpenJDK11U-jdk_aarch64_mac_hotspot_11.0.16.1_1.pkg` no error was shown but also no issues were found in the code. So I cloned the repo, bumped `npm-groovy-lint` to `11.1.1` and just copied the folder back to where the extension is supposed to be (`nicolasvuillamy.vscode-groovy-lint-2.0.0`) and voilà - it's working.

**Comment by ramiroachaval** [comment]
> Tried everything, even with `OpenJDK11U-jdk_aarch64_mac_hotspot_11.0.16.1_1.pkg` no error was shown but also no issues were found in the code. So I cloned the repo, bumped `npm-groovy-lint` to `11.1.1` and just copied the folder back to where the extension is supposed to be (`nicolasvuillamy.vscode-groovy-lint-2.0.0`) and voilà - it's working.
> 
> MacOS Ventura 13.5.2 (Apple Silicon) Node v20.6.1

Can you share the steps you made?

**Comment by kkreso** [comment]
> > Tried everything, even with `OpenJDK11U-jdk_aarch64_mac_hotspot_11.0.16.1_1.pkg` no error was shown but also no issues were found in the code. So I cloned the repo, bumped `npm-groovy-lint` to `11.1.1` and just copied the folder back to where the extension is supposed to be (`nicolasvuillamy.vscode-groovy-lint-2.0.0`) and voilà - it's working.
> > MacOS Ventura 13.5.2 (Apple Silicon) Node v20.6.1
> 
> Can you share the steps you made?

Well those are the steps, cloned this repo to a /tmp folder, bumped the npm-groovy-lint version in packages.json file, pulled the new package(s), ran build command and just copied folder from /tmp to the proper location.

**Comment by nvuillam** [comment]
@djukxe @kkreso @ramiroachaval  this is fixed in newest release v3.0.0, thanks to the great job of @stevenh :)

Please can you confirm it works well for your case now ?
- If yes, please rate the app -> https://marketplace.visualstudio.com/items?itemName=NicolasVuillamy.vscode-groovy-lint&ssr=false#review-details
- If not, please reopen the issue and we'll check :)
